allen-cell-animated/cell-feature-data
# Feature/process data on prod

Problem
=======
Closes #64 

Solution
========
What I/we did to solve this problem

- added a choice input for users to decide which db (stag or prod) they want to publish dataset to
- In `upload-dataset.yml`, added a second job that calls github api to get collaborators with admin access, and outputs all the admins for the next job to use 
- added a third job that will run `npm run process-dataset` with production DB. Only on `main` by admins


Now we have 2 manually triggered files in our workflows: 
1. `release-dataset.yml`: release data to staging or production, the job will only be triggered on `main` by authorized users for both databases
2. `upload-dataset.yml`: upload data to staging or production, anyone with write access to this repo should be able to publish data from any branches to **staging**, but only authorized users can publish data only from `main` branch to **production**     


## Type of change
Please delete options that are not relevant.

* New feature (non-breaking change which adds functionality)

* Repo admins can manage access [here](https://github.com/allen-cell-animated/cell-feature-data/settings/access)

## Steps to Verify
1. Go to Actions and choose `Upload dataset to staging or production DB` in workflows
2. Click on Run Workflow and select `feature/process-data-on-prod` branch for testing. 
    - Note that the job `process_dataset_to_production` is expected to be skipped on a non-main branch, comment out `if:  github.ref == 'refs/heads/main'` to run the job
4. Enter inputs and comment out `npm run process-dataset` for production DB testing
5. Run workflow

## Expected behavior 
- when DB == 'staging'
<img width="817" alt="Screenshot 2023-04-03 at 12 36 44 PM" src="https://user-images.githubusercontent.com/91452427/229609783-82506165-aed9-48e9-8dee-6a00e81cdc89.png">


- when DB == 'production' && authorized user && on `main`
<img width="827" alt="Screenshot 2023-04-03 at 12 36 58 PM" src="https://user-images.githubusercontent.com/91452427/229609737-de8bbdf0-12b6-4d8b-bf24-0b02f1245720.png">


- in the prod job, if not an admin, step`process dataset to production` will be skipped, step `Not authorized to process dataset to production` will run the echo statement 
<img width="1110" alt="Screenshot 2023-04-03 at 12 38 22 PM" src="https://user-images.githubusercontent.com/91452427/229610037-72101c4d-94d2-4694-87f9-792bf8733f3d.png">
